lib/matplotlib/cm.py

```diff
@@ -361,12 +361,19 @@ def to_rgba(self, x, alpha=None, bytes=False, norm=True):
 
     def set_array(self, A):
         """
-        Set the image array from numpy array *A*.
+        Set the image array from array-like *A*.
 
         Parameters
         ----------
-        A : ndarray or None
+        A : array-like or None
         """
+        A = cbook.safe_masked_invalid(A, copy=True) if A is not None else None
+
+        if (A is not None and A.dtype != np.uint8 and
+                not np.can_cast(A.dtype, float, "same_kind")):
+            raise TypeError("Image data of dtype {} cannot be converted to "
+                            "float".format(A.dtype))
```

Comment:
```suggestion
        if A is None:
            self._A = None
            return
            
        A = cbook.safe_masked_invalid(A, copy=True)

        if not np.can_cast(A.dtype, float, "same_kind"):
            raise TypeError("Image data of dtype {} cannot be converted to "
                            "float".format(A.dtype))
```

1. Returning early for None is simpler.
2. `np.can_cast(np.uint8, float, "same_kind")` is True, so `A.dtype != np.uint8` is not needed.

Reply:
ah, makes sense 👍🏼